sentinelhub/evalscript.py

```diff
@@ -0,0 +1,117 @@
+            }}],
+            output: [{output_spec}]
+        }}
+    }}
+
+    function updateOutputMetadata(scenes, inputMetadata, outputMetadata) {{
+        outputMetadata.userData = {{
+            "norm_factor":  inputMetadata.normalizationFactor
+        }}
+    }}
+
+    function evaluatePixel(sample) {{
+        return {{ {return_spec} }};
+    }}
+"""
+
+
+def parse_data_collection_bands(data_collection: DataCollection, bands: List[str]) -> List[Band]:
+    """Checks that all requested bands are available and returns the band information for further processing
+
+    :param data_collection: A collection of requested satellite data.
+    :param bands: A list of band or meta band names to use in the evalscript.
+    """
+
+    requested_bands = []
+    band_info_dict = {band_info.name: band_info for band_info in data_collection.bands + data_collection.metabands}
+
+    for band_name in bands:
+        if band_name not in band_info_dict:
+            raise ValueError(
+                f"Data collection {data_collection} does not have specifications for band {band_name}.\n"
+                f"Available bands are:\n{[band.name for band in data_collection.bands]}\nand meta-bands:\n"
+                f"{[band.name for band in data_collection.metabands]}."
+            )
+        requested_bands.append(band_info_dict[band_name])
+
+    return requested_bands
+
+
+def generate_evalscript(
+    data_collection: DataCollection,
+    bands: Optional[List[str]] = None,
+    meta_bands: Optional[List[str]] = None,
+    merged_output: Optional[str] = None,
+    use_dn: bool = True,
+) -> str:
+    """Generate an evalscript based on the provided specifications. This utility supports generating only evalscripts
+    with the mosaicking option set to `SIMPLE`.
+
+    :param data_collection: A collection of requested satellite data.
+    :param bands: A list of band names to use in the evalscript. Defaults to using all bands provided by the collection.
+    :param meta_bands: A list of meta band names to use in the evalscript. By default no meta bands are added.
+    :param merged_output: If provided, bands will be concatenated into a single multi-band tiff with this name.
+    :param use_dn: Use DN units if possible. Default is True. If DN units are not available, the default units for each
+        specific band are used. DN units will be used regardless of the flag if they are the only possible choice.
+    """
+
+    band_names = bands if bands is not None else [band.name for band in data_collection.bands]
+    meta_band_names = meta_bands if meta_bands is not None else []
+
+    input_names, input_units, output_spec, return_spec = [], [], [], []
+    requested_bands = parse_data_collection_bands(data_collection, band_names + meta_band_names)
+    for band in requested_bands:
+        unit_choice = band.units.index(Unit.DN) if (use_dn and Unit.DN in band.units) else 0
+        sample_type = DTYPE_TO_SAMPLE_TYPE[band.output_types[unit_choice]]
```

Comment:
this is the kind of bullshit that we could avoid by developing these evalscripts together with sh-py

Reply:
yep, felt like BS, but I guess it is the way it is.